staking: only disable slashed validators and keep then disabled for whole era

andresilva opened the pull request:
This PR changes the logic for disabling validators and trigger a new era. Whenever a validator commits an offence it is now tracked in the staking pallet and whenever a threshold is reached a new era is started. Only offences that led to a slash lead to the validator being disabled (e.g. if the validator is offline just by itself it is not disabled). Additionally the staking pallet makes sure that disabled validators stay disabled for the whole era.

The reason to not disable validators when there's no slash is https://github.com/paritytech/substrate/pull/9414 which will make sure that disabled validators can't author new blocks. If a validator is not slashed we can assume that the offence was not critical enough to block the validator from participating in consensus (it will be kicked from the set whenever the next era starts).

Fixes https://github.com/paritytech/polkadot/issues/2088.

andresilva commented on frame/staking/src/tests.rs:
```diff
@@ -2291,10 +2291,11 @@ fn slash_in_old_span_does_not_deselect() {
-		assert_eq!(Staking::force_era(), Forcing::NotForcing);
-		assert!(<Validators<Test>>::contains_key(11));
-		assert!(Session::validators().contains(&11));
+		// the validator doesn't get chilled again
+		assert!(<Staking as Store>::Validators::iter().find(|(stash, _)| *stash == 11).is_some());
+
+		// but we are still forcing a new era
+		assert_eq!(Staking::force_era(), Forcing::ForceNew);
```
Please check that this makes sense, I changed the logic so that a validator still gets disabled when there's an offence but it has a new slashing span. It will not get chilled though which was the original behavior.

shawntabrizi commented on frame/session/src/lib.rs:
```diff
@@ -663,42 +662,31 @@ impl<T: Config> Module<T> {
+			let i_u32 = i as u32;
+			if let Err(index) = disabled.binary_search(&i_u32) {
+				disabled.insert(index, i_u32);
+				T::SessionHandler::on_disabled(i);
```
it feels this `on_disabled` API should input a u32 not usize

shawntabrizi commented on frame/staking/src/lib.rs:
```diff
@@ -631,14 +630,18 @@ where
 		<pallet_session::Pallet<T>>::validators()
 	}
 
+	fn validators_len() -> usize {
```
this api should also be u32 not usize

shawntabrizi commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -433,6 +437,17 @@ pub mod pallet {
+	/// `OffendingValidatorsThreshold` is reached. The set is cleared when the era ends.
+	#[pallet::storage]
+	#[pallet::getter(fn offending_validators)]
+	pub type OffendingValidators<T: Config> = StorageValue<_, Vec<(u32, bool)>, ValueQuery>;
```
```suggestion
	pub type OffendingValidators<T: Config> = StorageValue<_, BoundedVec<(u32, bool)>, ValueQuery>;
```

Can you already make this a bounded vec? It will be something we will need to update and break in the future otherwise.

shawntabrizi replied:
note you will also need to add a bounding condition here

andresilva replied:
The bounding condition would be the maximum number of validators but that's dynamic (`MaxValidatorsCount` storage item), so I'm not sure how I could go about using it here.

shawntabrizi replied:
We should have a static upper limit which even bounds the dynamic limit, and then if we need to go higher, we need to do a runtime upgrade.

Unfortunately, I think this is the only way...

thiolliere replied:
yeah probably using a weak bonded vec is the easiest. I see we are going to have MaxAuthorities in pallet-authority-discovery, pallet-aura. I guess we can have same in pallet-staking. And use this max to bound this storage.

Anyway we can do in a follow-up

shawntabrizi commented on frame/staking/src/slashing.rs:
```diff
@@ -315,13 +313,54 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
+				offending.insert(index, (validator_index_u32, disable));
+
+				let offending_threshold = T::OffendingValidatorsThreshold::get() *
+					T::SessionInterface::validators_len() as u32;
```
i dont see why you are calling this again when you already call `validators()` above. looks like it can be optimized for one less read into a vec

shawntabrizi commented on frame/staking/src/slashing.rs:
```diff
@@ -315,13 +313,54 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
+
+				if offending.len() >= offending_threshold as usize {
+					// force a new era, to select a new validator set
+					<Pallet<T>>::ensure_new_era()
```
note to follow up here to understand if committing the changes to `offending` is needed in this scenario (as it is doing right now)

kianenigma replied:
This might have implications for the election provider. The test case `estimate_next_election_works` should be enough to ensure forcing a new era will not cause any issues.

thiolliere replied:
I think it is needed.

forcing a new era doesn't mean that next active session is the first session of a new era.
pallet_session is "delaying" session by one. So the when staking is configured with ForceNewEra then when pallet_session rotate session it will end the session (let's say) 3 (with `end_session(3)`), start the session 4 (with `start_session(4)`) and plan the session 5 (with `new_session(5)`) the session 5 will be the first session of the new era.

So we still need to disable the validators in the `start_session(4)`

shawntabrizi commented on frame/staking/src/slashing.rs:
```diff
@@ -315,13 +313,54 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
+				}
+			},
+			Ok(index) if disable => {
+				// the validator had previously offended without being disabled,
+				// let's make sure we disable it now
```
You want to put this comment under the if statement?

The validator is not guaranteed to not already be disabled

shawntabrizi commented on frame/staking/src/slashing.rs:
```diff
@@ -315,13 +313,54 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
+					T::SessionInterface::disable_validator(validator_index);
+				}
+			},
+			_ => {},
```
nit, i prefer not having catch-alls, and probably could be fixed by just moving the `disabled` if statement above into the match logic.

Same number of lines, but no catch all

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -605,14 +605,13 @@ pub struct UnappliedSlash<AccountId, Balance: HasCompact> {
-	/// disabled and block production stalls.
-	fn disable_validator(validator: &AccountId) -> Result<bool, ()>;
+	/// Disable the validator at the given index, returns `false` if the validator was already
+	/// disabled.
```
The index is assumed to never be out of bound?

andresilva replied:
I just added code to check if the index is out of bounds and returning `false` right away. Let me know if you think we should return `Result` instead.

kianenigma commented on frame/staking/src/mock.rs:
```diff
@@ -63,23 +59,14 @@ impl OneSessionHandler<AccountId> for OtherSessionHandler {
 	}
 
-	fn on_new_session<'a, I: 'a>(_: bool, validators: I, _: I)
+	fn on_new_session<'a, I: 'a>(_: bool, _validators: I, _: I)
```
```suggestion
	fn on_new_session<'a, I: 'a>(_: bool, _: I, _: I)
```

for consistency with the other arguments

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -359,6 +366,9 @@ impl<T: Config> Pallet<T> {
 			T::RewardRemainder::on_unbalanced(T::Currency::issue(rest));
+
+			// Clear offending validators.
+			<OffendingValidators<T>>::take();
```
```suggestion
			<OffendingValidators<T>>::kill();
```

or 

```suggestion
			let _ = <OffendingValidators<T>>::take();
```

shawntabrizi replied:
should kill for sure if the value is not used. should be more efficient

kianenigma reviewed (commented):
looks correct to me, will have a final review once comments are resolved.

emostov commented on frame/session/src/tests.rs:
```diff
@@ -336,7 +336,7 @@ fn session_keys_generate_output_works_as_set_keys_input() {
 
 #[test]
-fn return_true_if_more_than_third_is_disabled() {
+fn disable_index_returns_true_if_already_disabled() {
```
Do you "returns_false_if_already_disabled" (and true if not disabled and in the set) ?

andresilva replied:
Indeed, good catch.

emostov commented on frame/session/src/tests.rs:
```diff
@@ -345,10 +345,9 @@ fn return_true_if_more_than_third_is_disabled() {
 		force_new_session();
 		initialize_block(2);
 
+		assert_eq!(Session::disable_index(0), true);
```
Its not obvious to me why 0 would return true here. The way I read this 0 should not be a validator thus we would always expect false

andresilva replied:
The API takes a validator index and not the validator id, it's just a bit confusing because in the test environment the validator id are `usize`. So in this case `0` refers to validator id `1` above.

emostov commented on bin/node/runtime/src/lib.rs:
```diff
@@ -482,6 +477,7 @@ parameter_types! {
 	pub const SlashDeferDuration: pallet_staking::EraIndex = 24 * 7; // 1/4 the bonding duration.
 	pub const RewardCurve: &'static PiecewiseLinear<'static> = &REWARD_CURVE;
 	pub const MaxNominatorRewardedPerValidator: u32 = 256;
+	pub const OffendingValidatorsThreshold: Perbill = Perbill::from_percent(17);
```
Just out of curiosity, how do we come up with this number? And why not ~33?

andresilva replied:
By 33% finality would already be stalled so it would be a bit too late to take action. I didn't pick this number myself but I think the logic is to act when we're halfway through stalling consensus.

emostov commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -433,6 +437,17 @@ pub mod pallet {
 	#[pallet::getter(fn current_planned_session)]
 	pub type CurrentPlannedSession<T> = StorageValue<_, SessionIndex, ValueQuery>;
 
+	/// Indices of validators that have offended in the current era and whether they are currently
```
active era instead of current era?

emostov commented on frame/staking/src/mock.rs:
```diff
@@ -88,7 +75,12 @@ impl sp_runtime::BoundToRuntimeAppPublic for OtherSessionHandler {
+		None => return false,
+	};
+
+	Session::disabled_validators().contains(&validator_index)
```
Is this the only place we read `DisabledValidators`? I think I am missing the purpose of this storage item since afaict we keep it updated, but we only use it in this testing helper function.

emostov commented on frame/session/src/lib.rs:
```diff
@@ -663,42 +662,31 @@ impl<T: Config> Module<T> {
-				(true, disabled.len() as u32 > threshold)
-			} else {
-				(false, false)
+		DisabledValidators::mutate(|disabled| {
```
To clarify my understanding, is the purpose of the `DisabledValidators` storage item just to prevent unnescary calls to `T::SessionHandler::on_disabled(i)` ?

emostov replied:
I think I see: it will be used in babe/aura like https://github.com/paritytech/substrate/pull/9414/files#diff-f5430be578eba5004ae7f2f75f95f5ce7a2819be7babb0f7f0477cd027b35facR896-R900

andresilva replied:
I'm not against using a set here but I just kept the original data structure and we're using binary search to search this vec.

andresilva commented:
I think I've addressed all issues except bounding the disabled validators vec which I'm not sure how to do since there's no static bound right now.

emostov reviewed (approved):
Code LGTM with the caveat that I am new to a lot of this code so I may be missing broader implications outside of the diff.

Noting that it looks like we could have some minor optimizations  by making `DisabledValidators` a set as we mentioned here (https://github.com/paritytech/substrate/pull/9448/files#r692841892)

thiolliere commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -431,6 +435,17 @@ pub mod pallet {
+	/// This value should be a superset of disabled validators since not all offences lead to the
+	/// validator being disabled (if there was no slash). This is needed to track the percentage
+	/// of validators that have offended in the current era, ensuring a new era is forced if
+	/// `OffendingValidatorsThreshold` is reached. The set is cleared when the era ends.
```
maybe a quick note saying that the vec is sorted by the validator id, to avoid mistake in the future.

thiolliere reviewed (approved):
looks good to me

kianenigma commented:
@andresilva what's the plan for this?

andresilva commented:
@kianenigma Sorry, I was out for the past 2 weeks 😊 I've updated the PR with current master and added comments regarding the sorted vecs. I think I've addressed all comments from review except the usage of `BoundedVec`, I created https://github.com/paritytech/substrate/issues/9724 for this.

andresilva commented:
After merging this we need to make sure the polkadot companion isn't merged right away. I needed to point to a temporary beefy branch.

andresilva commented:
After talking to @kianenigma we decided to upgrade this to `needsaudit` since it's touching both session and staking. We'll wait until it's audited before merging.

andresilva commented:
bot merge

andresilva commented:
@ordian Thanks for updating the PR!
